classes/task/merge_user_accounts.php

```diff
@@ -0,0 +1,198 @@
+ * @author    Nicola Vallinoto <n.vallinoto@liguriadigitale.it>
+ * @copyright 2013 Servei de Recursos Educatius (http://www.sre.urv.cat)
+ * @copyright 2023 Liguria Digitale (http://www.liguriadigitale.it)
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ * */
+class merge_user_accounts extends \core\task\adhoc_task {
+    /**
+     * Execute the task of merge users accounts.
+     */
+    public function execute() {
+        global $DB;
+        $data = $this->get_custom_data();
+        $maxattempts = get_config('tool_mergeusers', 'maxattempts');
+        $mergerequest = $DB->get_record(merge_request::TABLE_MERGE_REQUEST,
+                                            ['id' => $data->mergerequestid]);
+        if (!$mergerequest) {
+            // Merge request was removed before running this task.
+            return;
+        }
+        $this->verify_users_to_keep_and_remove($mergerequest);
+        $mergerequestresult = $this->merge($mergerequest, $maxattempts, merge_request::TRIED_WITH_ERROR);
+        if ($mergerequestresult->status == merge_request::COMPLETED_WITH_SUCCESS) {
+            /* We run the merge request AGAIN because the user may be interacting with Moodle
+            * while merge request is being processed, so that NO ALL records are correctly migrated
+            * into the user to keep. It will ensure ALL records are correctly migrated into the user to keep. */
+            $mergerequestresult = $this->merge($mergerequestresult, $maxattempts, merge_request::COMPLETED_WITH_ERRORS);
+        }
+        if ($mergerequestresult->status != merge_request::COMPLETED_WITH_ERRORS &&
+            $mergerequestresult->status != merge_request::COMPLETED_WITH_SUCCESS) {
+            // Throwing exception will ensure this adhoc task is re-queued until $maxretries is reached.
+            throw new moodle_exception(get_string('failedmergerequest', 'tool_mergeusers'));
+        }
+    }
+    /**
+     * Function for merging two users.
+     */
+    private function merge(object $record, int $maxattempts, int $statusforerror) {
+        $retries = $record->retries + 1;
+        // Update retries.
+        $this->update_retries_in_table($record->id, $retries);
+        $logs = [];
+        $logs = $record->log;
+        $mut = new MergeUserTool();
+        list($success, $log) = $mut->merge($record->keepuserid, $record->removeuserid);
+        if (!$success) {
+            if ($retries >= $maxattempts) {
+                $status = merge_request::COMPLETED_WITH_ERRORS;
+                // Send a notification to administrator ?
+            } else {
+                $status = $statusforerror;
+            }
+        } else {
+            $status = merge_request::COMPLETED_WITH_SUCCESS;
+        }
+        // Append logs to the list.
+        // $baseitem->log[$item->timemodified] = json_decode($item->log, false);
+        $logs[$record->timeadded] = $log;
+        $this->update_status_and_log_in_table($record->id,
```

Comment:
I think we should add a new method to the merge_request class, like this:

```php
public function get_log_as_json(): string {
    return $this->data->log;
}
```

And update the `__set()`  method to consider this, by adding this piece of code:

```php
        if ($name == 'status && 
           ($value == merge_request::COMPLETED_WITH_SUCCESS ||
            $value == merge_request::COMPLETED_WITH_ERRORS)) {
                $this->timecompleted = time();
        }
```

Then, to simplify the operation of the update_status_and_log method, by having this:

```php
private function update_status_and_log(merge_request $mergerequest): void {
    ...
    $update = (object)[
        'id' => $mergerequest->id,
       'status' => $mergerequest->status,
       'timecompleted' => $mergerequest->timecompleted,
      'log' => $mergerequest->get_log_as_json(),
    ];
    $DB->....
}
```

so we fully delegate on the merge_request class to update any field according to the business logic, internally, and not externally.

Also, the 2 lines have to be in the reverse order:

```php
$mergerequest->status = $status;
$this->update_status_and_log($mergerequest);
```

So the status is up-to-date when updating it inside the update_status_and_log method.